### PR TITLE
Python23support

### LINKED ISSUE #68: python 3 support
Currently code doesn't work with python 3.

### PR DESCRIPTION
**Pull request by twerkmeister**
Made the codebase python 2 and 3 compatible (Fixes #68 ) . I prioritized the python 3 convention in case of differences and added the following to the documentation:

```
To ensure cross compatibility between Python 2 and 3 we prioritize Python 3 conventions.
Keep in mind that:

- all string literals are unicode strings
- division generates floating point numbers. Use ``//`` for truncated division
- some built-ins, e.g. ``map`` and ``filter`` return iterators in Python 3. 
  If you want to make use of them import the Python 3 version of them from ``builtins``. 
  Otherwise use list comprehensions, which work uniformly across versions
- use ``io.open`` instead of the builtin ``open`` when working with files
```

### CODE REVIEW

**Comment by twerkmeister** [comment]
Lately a lot of the builds fail while downloading spacy ... 
```
$ pip install https://github.com/explosion/spacy-models/releases/download/de_core_news_md-1.0.0/de_core_news_md-1.0.0.tar.gz --no-cache-dir > jnk
Exception:
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/pip/commands/install.py", line 324, in run
    requirement_set.prepare_files(finder)
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/pip/req/req_set.py", line 380, in prepare_files
    ignore_dependencies=self.ignore_dependencies))
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/pip/req/req_set.py", line 620, in _prepare_file
    session=self.session, hashes=hashes)
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/pip/download.py", line 821, in unpack_url
    hashes=hashes
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/pip/download.py", line 659, in unpack_http_url
    hashes)
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/pip/download.py", line 882, in _download_http_url
    _download_url(resp, link, content_file, hashes)
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/pip/download.py", line 605, in _download_url
    consume(downloaded_chunks)
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/pip/utils/__init__.py", line 852, in consume
    deque(iterator, maxlen=0)
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/pip/download.py", line 571, in written_chunks
    for chunk in chunks:
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/pip/utils/ui.py", line 139, in iter
    for x in it:
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/pip/download.py", line 560, in resp_read
    decode_content=False):
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/pip/_vendor/requests/packages/urllib3/response.py", line 357, in stream
    data = self.read(amt=amt, decode_content=decode_content)
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/pip/_vendor/requests/packages/urllib3/response.py", line 324, in read
    flush_decoder = True
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/contextlib.py", line 35, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/pip/_vendor/requests/packages/urllib3/response.py", line 246, in _error_catcher
    raise ReadTimeoutError(self._pool, None, 'Read timed out.')
ReadTimeoutError: HTTPSConnectionPool(host='github-cloud.s3.amazonaws.com', port=443): Read timed out.
The command "pip install https://github.com/explosion/spacy-models/releases/download/de_core_news_md-1.0.0/de_core_news_md-1.0.0.tar.gz --no-cache-dir > jnk" failed and exited with 2 during .
Your build has been stopped.
```

File: rasa_nlu/config.py [hunk 1]
```diff
@@ -70,7 +75,7 @@ def __init__(self, filename=None, env_vars=None, cmdline_args=None):
                 warnings.warn("No pipeline specified and unknown pipeline template " +
                               "'{}' passed.".format(self.__dict__['pipeline']))
 
-        for key, value in self.items():
+        for key, value in list(self.items()):
```
Comment (amn41) [inline rasa_nlu/config.py]: isn't this `list` redundant if `self.items()` already casts as a list?

File: rasa_nlu/config.py [hunk 2]
```diff
@@ -89,13 +94,13 @@ def __len__(self):
...
     def items(self):
-        return self.__dict__.items()
+        return list(self.__dict__.items())
 
     def view(self):
         return json.dumps(self.__dict__, indent=4)
 
     def format_env_vars(self, env_vars):
-        keys = [key for key in env_vars.keys() if "RASA" in key]
+        keys = [key for key in list(env_vars.keys()) if "RASA" in key]
```
Comment (amn41) [inline rasa_nlu/config.py]: just a question: in py3 .keys() and .items() return iterators, right? so why do we have to convert to a list if we're in a list comprehension?

File: rasa_nlu/featurizers/mitie_featurizer.py [hunk 3]
```diff
@@ -45,7 +50,7 @@ def features_for_tokens(self, tokens, feature_extractor):
         for token in tokens:
             vec += feature_extractor.get_feature_vector(token)
         if tokens:
-            return vec / len(tokens)
+            return old_div(vec, len(tokens))
```
Comment (amn41) [inline rasa_nlu/featurizers/mitie_featurizer.py]: vec is already an array of floats, right? so we don't need old_div, do we?

File: rasa_nlu/featurizers/spacy_featurizer.py [hunk 4]
```diff
@@ -48,7 +53,7 @@ def features_for_doc(self, doc, nlp):
             if token.has_vector:
                 vec.append(token.vector)
         if vec:
-            return np.sum(vec, axis=0) / len(vec)
+            return old_div(np.sum(vec, axis=0), len(vec))
```
Comment (amn41) [inline rasa_nlu/featurizers/spacy_featurizer.py]: same here-  not sure we need old_div

File: rasa_nlu/version.py [hunk 5]
```diff
@@ -1 +1,6 @@
+from __future__ import unicode_literals
+from __future__ import print_function
+from __future__ import division
+from __future__ import absolute_import
+
```
Comment (amn41) [inline rasa_nlu/version.py]: worth a version bump I think :)

**Comment by amn41** [comment]
re the failed builds, if it's just *some* of the builds, and looking at the HTTP error, I agree it's likely that it's a network thing. I know we already run the tests on travis one at a time (to prevent CI server running out of memory). Are different python versions run on different VMs? Github's S3 getting overloaded by a few simultaneous downloads would be surprising, no? Maybe we're limited by travis's network?

File: docs/contribute.rst [hunk 6]
```diff
@@ -10,7 +10,19 @@ We created a tag that should get you started quickly if you are searching for
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To ensure cross compatibility between Python 2 and 3 we prioritize Python 3 conventions.
+Keep in mind that:
+
+- all string literals are unicode strings
+- division generates floating point numbers. Use ``//`` for truncated division
+- some built-ins, e.g. ``map`` and ``filter`` return iterators in Python 3. If you want to make use of them import the Python 3 version of them from ``builtins``. Otherwise use list comprehensions, which work uniformly across versions
+- use ``io.open`` instead of the builtin ``open`` when working with files
+
```
Comment (tmbo) [inline docs/contribute.rst]: pls also add the necessary or suggested imports for each file

File: rasa_nlu/emulators/api.py [hunk 7]
```diff
@@ -11,12 +16,12 @@ def __init__(self):
...
     def normalise_response_json(self, data):
         # populate entities dict
-        entities = {entity_type: [] for entity_type in set(map(lambda x: x["entity"], data["entities"]))}
+        entities = {entity_type: [] for entity_type in set([x["entity"] for x in data["entities"]])}
         for entity in data["entities"]:
             entities[entity["entity"]].append(entity["value"])
 
         return {
-            "id": unicode(uuid.uuid1()),
+            "id": str(uuid.uuid1()),
```
Comment (tmbo) [inline rasa_nlu/emulators/api.py]: why do we need the `str` here?

File: rasa_nlu/config.py (hunk 1 above)
Comment (twerkmeister) [inline rasa_nlu/config.py]: in this specific case it's unnecessary true. Was part of the automatic conversion that just ensured exactly uniform return types. Similar for some of the other things you highlighted. Slipped my mind to check some of the automatically converted parts in detail.

File: rasa_nlu/featurizers/mitie_featurizer.py (hunk 3 above)
Comment (twerkmeister) [inline rasa_nlu/featurizers/mitie_featurizer.py]: In this specfic case we can use `/`. With `//` we would get a truncated float, so `//` and `old_div` are not exactly the same

File: rasa_nlu/emulators/api.py (hunk 7 above)
Comment (twerkmeister) [inline rasa_nlu/emulators/api.py]: str is here comes from `from builtins import str` which is a backport of the python 3 str class and subclassing the python2 unicode class. This way its behaving similar as before.

**Comment by twerkmeister** [comment]
Ok I fixed the mentioned things. Approved?

**Pull request merged**